### PR TITLE
BFD-3006: Prepare to sunset bfd-public-test-data bucket

### PR DESCRIPTION
**Pull request by rcrupp**
**JIRA Ticket:**
[BFD-3006](https://jira.cms.gov/browse/BFD-3006)

**User Story or Bug Summary:**
As a BFD contributor, I want the ability to reliably access BFD public test data so that I can be empowered to explore the BFD ecosystem without unnecessary restrictions on the test data.

---

### What Does This PR Do?
Adds lifecycle policy to `bfd-public-test-data` bucket to move files to cold storage, and adds a new `bfd-test-bucket` to house the test data files.
* Adds a lifecycle configuration to the existing `bfd-public-test-data` bucket.
* Adds a new S3 bucket, `bfd-test-bucket` with a lifecycle configuration. 

### What Should Reviewers Watch For?
If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.
* Verify that the new file belongs in the `services/common` module or should it belong to an environment.
* The `bfd-public-test-data` bucket is not currently managed with terraform, so I had to import it.  Review the plan below to make sure it's not adding anything unnecessary to the bucket.
* I used the name `bfd-test-data` for the new bucket name, but if there's a better, more descriptive name that applies I'm happy to change it.

```
## `terraform plan` run in bfd-3006-test workspace:

Terraform will perform the following actions:

  # aws_s3_bucket.bfd-public-test-data will be updated in-place
  ~ resource "aws_s3_bucket" "bfd-public-test-data" {
      + force_destroy               = false
        id                          = "bfd-public-test-data"
        tags                        = {
            "cms-cloud-exempt:public-s3-bucket" = "https://confluence.cms.gov/x/VFNBGg"
        }
      ~ tags_all                    = {
          + "Environment"                       = "test"
          + "Terraform"                         = "true"
          + "stack"                             = "bfd-3006-test"
          + "tf_module_root"                    = "ops/terraform/services/common"
            # (3 unchanged elements hidden)
        }
        # (9 unchanged attributes hidden)

        # (4 unchanged blocks hidden)
    }

  # aws_s3_bucket.bfd-test-data will be created
  + resource "aws_s3_bucket" "bfd-test-data" {
      + acceleration_status         = (known after apply)
      + acl                         = (known after apply)
      + arn                         = (known after apply)
      + bucket                      = "bfd-test-data"
      + bucket_domain_name          = (known after apply)
      + bucket_prefix               = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + object_lock_enabled         = (known after apply)
      + policy                      = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = {
          + "Environment"    = "test"
          + "Terraform"      = "true"
          + "application"    = "bfd"
          + "business"       = "oeda"
          + "stack"          = "bfd-3006-test"
          + "tf_module_root" = "ops/terraform/services/common"
        }
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)
    }

  # aws_s3_bucket_lifecycle_configuration.bfd-public-test-data will be created
  + resource "aws_s3_bucket_lifecycle_configuration" "bfd-public-test-data" {
      + bucket = "bfd-public-test-data"
      + id     = (known after apply)

      + rule {
          + id     = "rule-1"
          + status = "Enabled"

          + filter {
            }

          + noncurrent_version_transition {
              + noncurrent_days = 60
              + storage_class   = "GLACIER"
            }
        }
    }

  # aws_s3_bucket_lifecycle_configuration.bfd-test-data will be created
  + resource "aws_s3_bucket_lifecycle_configuration" "bfd-test-data" {
      + bucket = (known after apply)
      + id     = (known after apply)

      + rule {
          + id     = "rule-1"
          + status = "Enabled"

          + filter {
            }

          + noncurrent_version_transition {
              + noncurrent_days = 60
              + storage_class   = "GLACIER"
            }
        }
    }
data.aws_security_group.vpn: Reading...
  # aws_s3_bucket_public_access_block.bfd-test-data will be created
  + resource "aws_s3_bucket_public_access_block" "bfd-test-data" {
      + block_public_acls       = true
      + block_public_policy     = true
      + bucket                  = (known after apply)
      + id                      = (known after apply)
      + ignore_public_acls      = true
      + restrict_public_buckets = true
    }
data.aws_security_group.management: Read complete after 1s [id=sg-04d406660d1da0d6b]
  # aws_s3_bucket_server_side_encryption_configuration.bfd-public-test-data will be imported
    resource "aws_s3_bucket_server_side_encryption_configuration" "bfd-public-test-data" {
        bucket = "bfd-public-test-data"
        id     = "bfd-public-test-data"
aws_s3_bucket_server_side_encryption_configuration.this: Refreshing state... [id=bfd-bfd-3006-test-admin20231101145708064800000002]
        rule {
            bucket_key_enabled = false

            apply_server_side_encryption_by_default {
                sse_algorithm = "AES256"
            }
        }
    }

  # aws_s3_bucket_server_side_encryption_configuration.bfd-test-data will be created
  + resource "aws_s3_bucket_server_side_encryption_configuration" "bfd-test-data" {
      + bucket = (known after apply)
      + id     = (known after apply)

      + rule {
          + apply_server_side_encryption_by_default {
              + sse_algorithm = "AES256"
            }
        }
    }

  # aws_s3_bucket_versioning.bfd-public-test-data will be imported
    resource "aws_s3_bucket_versioning" "bfd-public-test-data" {
        bucket = "bfd-public-test-data"
        id     = "bfd-public-test-data"

        versioning_configuration {
            status = "Enabled"
        }
    }

  # aws_s3_bucket_versioning.bfd-test-data will be created
  + resource "aws_s3_bucket_versioning" "bfd-test-data" {
      + bucket = (known after apply)
      + id     = (known after apply)

      + versioning_configuration {
          + mfa_delete = (known after apply)
          + status     = "Enabled"
        }
    }

Plan: 2 to import, 6 to add, 1 to change, 0 to destroy.
```

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist
I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: ops/terraform/services/common/s3-test-data.tf [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  bucket = aws_s3_bucket.bfd-public-test-data.id
+
+  rule {
+    id = "rule-1"
+
+    filter {}
+
+    noncurrent_version_transition {
+      noncurrent_days = 60
+      storage_class   = "GLACIER"
```
Comment (keithdadkins) [inline ops/terraform/services/common/s3-test-data.tf:55]: Maybe a better option here would be [s3 Intelligent Tiering](https://aws.amazon.com/s3/storage-classes/). This would automatically move objects to cheaper storage depending on usage. Ie, objects not used in 30 days get moved to a cheaper tier, same for 60, and cold storage after 90. Cost is similar to Glacier without the retrieval penalty. @mjburling - I was researching this earlier as it may be a good fit for most if not all of our buckets (eft, etl, etc).

Comment (keithdadkins) [reply ops/terraform/services/common/s3-test-data.tf:55]: $.10 per million objects fee for tracking object usage (https://aws.amazon.com/s3/pricing/?nc=sn&loc=4)

File: ops/terraform/services/common/s3-test-data.tf [hunk 2]
```diff
@@ -0,0 +1,108 @@
+
+data "aws_canonical_user_id" "current" {}
+
+# bfd-public-test-data S3 bucket imports, not currently managed with terraform
+import {
```
Comment (keithdadkins) [inline ops/terraform/services/common/s3-test-data.tf:5]: I wasn't even aware of this syntax. Seems straightforward enough, though there is the _potential_ this could break in the future:
> Experimental: While we do not expect to make backwards-incompatible changes to syntax, the -generate-config-out flag and how Terraform processes imports during the plan stage and generates configuration may change in future releases.

**Review by mjburling — CHANGES_REQUESTED** [review]
This needs a _little_ more guidance, I think.

### Definition Location
tl;dr; `bfd-public-test-data` should probably be defined under `ops/terraform/env/mgmt`

The _terraservices_ under `ops/terraform/services` are SDLC-environment scoped, i.e. resources specific to a given environment are defined and managed here. For resources like `bfd-public-test-data` that are **not** specific to any one SDLC-environment, we generally define these in `mgmt`. Note that historically, _some_ resources like this had been defined under `ops/terraform/env/global/s3`, but there's an effort to re-locate these resources in the near term.

### Usage of `import`
tl;dr; Import the resources however you'd like, but don't keep the ephemeral `import` blocks if that's the route you choose.

Thanks for working on importing `bfd-public-test-data`–I _thought_ we did this last year!

I probably would have imported these resources one-by-one into the appropriate state, though I do see the value in the import blocks for an initial import of... say three or more resources as you have here. However, those blocks are strictly optional once the resources are in state and have accurate, complementary resource blocks. I don't think we want or need the ephemeral import blocks as reminders of the historical ClickOps configuration practices, etc.

### Lifecycle Policies
tl;dr; Defining lifecycle might be premature. We might consider skipping it for now.

I think there was another story to evaluate how peering partners or their customers may (or may not be) using the test data that's defined here. The assumption now is that this data has been generally unsuitable for the past 12-18 months of BFD's development, but... maybe there's a use-case that we're not thinking of just yet?

It's funny to think that this is premature, given that this bucket has been in-use for so long, but... it's possible that these resources are not used for any of the use-cases we intend. Instead, what we see in terms of resources accessed have been security researchers/penetration testers landing on our publicly available data. That said, we might just skip the life cycle policies all together, package what data exists into a single archive, stash it as a single object elsewhere, and delete this bucket.

**Comment by rcrupp** [comment]
Ok, I can move it under /env/mgmt, that makes sense, and I'll take out the lifecycle config.  From the conversation on Slack in #bfd-users referenced in [BFD-3007](https://jira.cms.gov/browse/BFD-3007) it looks like none of the peering partners use it.  Only mentions I found were on Slack and in some disabled tests.  So does anything other than importing the bucket need to be done?  Do we need a new bucket at all?  Can I remove the code that references the bucket as part of this story?

**Comment by mjburling** [comment]
> Ok, I can move it under /env/mgmt, that makes sense, and I'll take out the lifecycle config. From the conversation on Slack in #bfd-users referenced in [BFD-3007](https://jira.cms.gov/browse/BFD-3007) it looks like none of the peering partners use it. Only mentions I found were on Slack and in some disabled tests. So does anything other than importing the bucket need to be done? Do we need a new bucket at all? Can I remove the code that references the bucket as part of this story?

I've _just_ reviewed BFD-3007. With the outcomes there in mind, I'll soften my position on the lifecycle policies and recommend the following.
1. Confer with @keithdadkins on subsequent items
2. Import the bucket
3. Maintain the lifecycle policies with a move to cold storage soonest
4. Remove public ACLs
5. Adjust `ops/terraform/env/mgmt/s3-account.tf` to enable public access blocks, address comments therein
6. Ensure references to this bucket in documentation in the whole of this project (gh repository, gh wiki, and... maybe confluence?) are eliminated _or_ annotated to indicate these resources are no longer publicly available, fit for general use, etc.

**Review by lsmitchell — APPROVED** [review]
Seems good to me from the application side; we don't use the stored 'synthetic' data or tests so might as well delete it.

**Review by brandoncruz3 — DISMISSED** [review]

**Comment by rcrupp** [comment]
Updated the s3 bucket config and tested in mgmt by importing the bucket and running terraform plan.  It will add 5 resources, update 2 and delete none per the plan.

**Review by keithdadkins — APPROVED** [review]
Looks good and thanks for seeing this through!

**Review by mackec1 — APPROVED** [review]

File: ops/terraform/env/mgmt/s3.tf [hunk 3]
```diff
@@ -119,3 +119,57 @@ resource "aws_s3_bucket_policy" "logging" {
...
+
+resource "aws_s3_bucket_ownership_controls" "bfd_public_test_data" {
+  bucket = aws_s3_bucket.bfd_public_test_data.id
+  rule {
+    object_ownership = "BucketOwnerEnforced"
+  }
+}
+
+resource "aws_s3_bucket_lifecycle_configuration" "bfd_public_test_data" {
+  depends_on = [aws_s3_bucket_versioning.bfd_public_test_data]
```
Comment (mjburling) [inline ops/terraform/env/mgmt/s3.tf:160]: Is this explicit `depend_on` necessary? I think the terraform DAG will take care of this, given the reference to the bucket by address just below, right?

File: ops/terraform/env/mgmt/s3.tf [hunk 4]
```diff
@@ -119,3 +119,57 @@ resource "aws_s3_bucket_policy" "logging" {
...
+    filter {}
+
+    transition {
+      days            = 7
+      storage_class   = "GLACIER_IR"
+    }
+
+    status = "Enabled"
+  }
+}
```
Comment (mjburling) [inline ops/terraform/env/mgmt/s3.tf:175]: Just a nit... could you adjust your editor to remove this _very_ mild warning about the absence of new lines at EOF?

https://thoughtbot.com/blog/no-newline-at-end-of-file#following-the-rules-in-your-editor

File: ops/terraform/env/mgmt/s3-account.tf [hunk 5]
```diff
@@ -1,11 +1,9 @@
...
-# Please note, that as long as we host public test data from the `bfd-public-test-data` s3 bucket,
-# we can only block public acls/policies on *new* s3 buckets.
 
 resource "aws_s3_account_public_access_block" "this" {
   block_public_acls       = true
   block_public_policy     = true
-  ignore_public_acls      = false # leave false due to bfd-public-test-data bucket
-  restrict_public_buckets = false # leave false due to bfd-public-test-data bucket
+  ignore_public_acls      = true
+  restrict_public_buckets = true
```
Comment (mjburling) [inline ops/terraform/env/mgmt/s3-account.tf:7]: ❤️ At last!

**Review by mjburling — CHANGES_REQUESTED** [review]
👋 Hey, Randall! Thanks for taking care of all of this. I just noted one or two minor things that are at the heart of this story. I also see that the scope has expanded somewhat, but it's consistent with the spirit of the story–I think it's a good idea! 

Admittedly, I'm less familiar with the history surrounding _that_ synthetic data, so I'll defer to others with more direct experience. That said, it looks like it's a popular decision 💯

**Review by lsmitchell — APPROVED** [review]
Good with the synthetic test data removal if the tf stuff is settled.

File: ops/terraform/env/mgmt/s3.tf (hunk 3 above)
Comment (rcrupp) [reply ops/terraform/env/mgmt/s3.tf]: it is not, I removed it.

**Review by mjburling — APPROVED** [review]

**Review by lsmitchell — APPROVED** [review]

**Review by brandoncruz3 — APPROVED** [review]
